libc/string/string_tokpbrkstr.c

```diff
@@ -0,0 +1,645 @@
+}
+
+
+TEST(string_tok, empty_input_output)
+{
+	char empty[] = { 0 };
+
+	/* Checking if strtok will read all elements in substring and another his search will return null*/
+	TEST_ASSERT_NOT_NULL(strtok("abc", "d"));
+	TEST_ASSERT_EQUAL_PTR(NULL, strtok(NULL, "d"));
+
+	/* Different scenarios of acquiring null pointer as output*/
+	TEST_ASSERT_EQUAL_PTR(NULL, strtok("", "d"));
+	TEST_ASSERT_EQUAL_PTR(NULL, strtok(NULL, ""));
+	TEST_ASSERT_EQUAL_PTR(NULL, strtok("", "d"));
+	TEST_ASSERT_EQUAL_PTR(NULL, strtok(NULL, "\0"));
+
+	TEST_ASSERT_EQUAL_PTR(NULL, strtok(empty, ""));
+	TEST_ASSERT_EQUAL_PTR(NULL, strtok(NULL, "\0"));
+	TEST_ASSERT_EQUAL_PTR(NULL, strtok(empty, ""));
+	TEST_ASSERT_EQUAL_PTR(NULL, strtok(NULL, ""));
+
+	TEST_ASSERT_EQUAL_PTR(NULL, strtok(empty, "\0"));
+	TEST_ASSERT_EQUAL_PTR(NULL, strtok(NULL, "\0"));
+	TEST_ASSERT_EQUAL_PTR(NULL, strtok(empty, "\0"));
+	TEST_ASSERT_EQUAL_PTR(NULL, strtok(NULL, ""));
+}
+
+
+TEST(string_tok, call_to_call_sep)
+{
+	char loremStr[20] = LOREM_IPSUM,
+		 separator[2] = { 0 };
+	int i;
+
+
+	/*
+	 * In this case we checking if is possible to
+	 * pass different stop points in each call to
+	 * get desired tokens
+	 */
+	TEST_ASSERT_EQUAL_PTR(loremStr, strtok(loremStr, "I"));
+	TEST_ASSERT_EQUAL_PTR(&loremStr[7], strtok(NULL, "m"));
+	TEST_ASSERT_EQUAL_PTR(&loremStr[11], strtok(NULL, "D"));
+	TEST_ASSERT_EQUAL_PTR(&loremStr[13], strtok(NULL, "r"));
+
+	/*
+	 * After using strtok on some elements we must
+	 * again write it to the array if we want to use
+	 * the same set because strtok breaks the elements in the string
+	 */
+	sprintf(loremStr, "%s", LOREM_IPSUM);
+
+	/*
+	 * After passing all delimiters we cannot avoid misplaced tokens because
+	 * strtok looks for all elements in "ImDr" and breaks a string
+	 * in points where the first occurrence of it is found
+	 */
+	TEST_ASSERT_EQUAL_PTR(loremStr, strtok(loremStr, "ImDr"));
+	TEST_ASSERT_EQUAL_PTR(&loremStr[3], strtok(NULL, "ImDr"));
+	TEST_ASSERT_EQUAL_PTR(&loremStr[5], strtok(NULL, "ImDr"));
+	TEST_ASSERT_EQUAL_PTR(&loremStr[7], strtok(NULL, "ImDr"));
+	TEST_ASSERT_EQUAL_PTR(&loremStr[11], strtok(NULL, "ImDr"));
+	TEST_ASSERT_EQUAL_PTR(&loremStr[13], strtok(NULL, "ImDr"));
+	TEST_ASSERT_EQUAL_PTR(NULL, strtok(NULL, "ImDr"));
+
+
+	create_ascii_set(1);
+
+	/*
+	 * In this case, we tokenize the whole ASCII set using jump-over by
+	 * 2 elements. We are forced to jump over 2 places because strtok set
+	 *  divider placed as a null term and in this case, we are unable to
+	 * cover the whole set. Jumping over 2 places guarantees us output in the form
+	 * of one element form ascii set to do an assert with the place where
+	 * strtok has saved the stopping place
+	 */
+	TEST_ASSERT_EQUAL_PTR(asciiStr, strtok(asciiStr, "\2"));
+	for (i = 4; i < sizeof(asciiStr) - 1; i += 2) {
+		separator[0] = i;
+		TEST_ASSERT_EQUAL_PTR(&asciiStr[i - 2], strtok(NULL, &separator[0]));
+	}
+}
```

Comment:
We could also add a check, where separator is a character from the beginning part of a string, which shouldn't be available in next calls, so for example, when you call:

```
char str[] = "abc";
TEST_ASSERT_EQUAL_PTR(str, strtok(str, "b"));
```
you want to:
```
TEST_ASSERT_EQUAL_PTR(&str[2], strtok(str, "a"));
```

Reply:
Those cases are checked with previous checks